File: src/main/java/pro/tuscan/config/properties/FaceitWebClientProperties.java
package pro.tuscan.config.properties;

import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.context.annotation.Configuration;
import org.springframework.validation.annotation.Validated;

import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;

@Validated
@Configuration
@ConfigurationProperties(prefix = "faceit")
public class FaceitWebClientProperties {

    @NotEmpty
    private String url;

    @NotEmpty
    private String openUrl;

    @NotNull
    private Timeout timeout;

    @NotEmpty
    private String apiKey;

    @NotNull
    private Retry retry;

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public Timeout getTimeout() {
        return timeout;
    }

    public void setTimeout(Timeout timeout) {
        this.timeout = timeout;
    }

    public String getApiKey() {
        return apiKey;
    }

    public void setApiKey(String apiKey) {
        this.apiKey = apiKey;
    }

    public String getOpenUrl() {
        return openUrl;
    }

    public void setOpenUrl(String openUrl) {
        this.openUrl = openUrl;
    }

    public Retry getRetry() {
        return retry;
    }

    public void setRetry(Retry retry) {
        this.retry = retry;
    }
}

File: src/main/java/pro/tuscan/config/properties/FaceitWebClientProperties.kt
package pro.tuscan.config.properties

import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.context.properties.ConstructorBinding
import org.springframework.validation.annotation.Validated
import javax.validation.constraints.Min
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.NotNull

@Validated
@ConfigurationProperties(prefix = "faceit")
@ConstructorBinding
data class FaceitWebClientProperties(
        @NotEmpty val url: String,
        @NotEmpty val openUrl: String,
        @NotNull val timeout: Timeout,
        @NotEmpty val apiKey: String,
        @NotNull val retry: Retry
)

data class Retry(
        @Min(1) val backoff: Long,
        @Min(1) val maxRetries: Long
)

data class Timeout(
        @Min(1) val read: Long,
        @Min(1) val connect: Long
)

